server, lint: centralize and restrict HTTP cookie construction

dhartunian opened the pull request:
### server: centralize cookie construction

All cookie construction for session and tenant cookies is now moved to
explicit constructors instead of initializing `http.Cookie{}` structs
directly. This reduces the chance of error and explicitly requires
setting the `secure` flag via the constructor.

The constructs were created to exactly mimic all existing cookie
settings. No change in flags should happen as a result of this commit.

A separate commit will introduce a linter to prevent raw cookie
creation.

Epic: None
Release note: None

### lint: add linter to reject raw http.Cookie structs

We previously introduced bugs when creating HTTP cookies in new code
that did not have the proper fields set. This linter aims to enforce
use of cookies only through pre-generated constructors in `pkg/server/
authserver/cookie.go`. This will reduce the chance of errors in the
future and keep things consistent.

Epic: None
Release note: None

dhartunian committed:
server: centralize cookie construction

All cookie construction for session and tenant cookies is now moved to
explicit constructors instead of initializing `http.Cookie{}` structs
directly. This reduces the chance of error and explicitly requires
setting the `secure` flag via the constructor.

The constructs were created to exactly mimic all existing cookie
settings. No change in flags should happen as a result of this commit.

A separate commit will introduce a linter to prevent raw cookie
creation.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119748)
<!-- Reviewable:end -->

dhartunian commented:
Please only review last 2 commits. First two getting merged here: https://github.com/cockroachdb/cockroach/pull/119261

stevendanna commented on pkg/server/authserver/cookie.go:
```diff
@@ -174,3 +175,55 @@ func findTenantSelectCookieValue(cookies []*http.Cookie) string {
 	return ""
 }
+
+func CreateSessionCookie(value string, forHTTPSOnly bool) *http.Cookie {
```
Sorry to be a pain: Do we want some function-level comments on these functions?

stevendanna commented on pkg/testutils/lint/passes/norawcookies/norawcookies.go:
```diff
@@ -0,0 +1,48 @@
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name:     "norawcookies",
+	Doc:      `check for correct use of Cookie constructors`,
+	Requires: []*analysis.Analyzer{inspect.Analyzer},
+	Run:      run,
+}
```
You've already written it so I wouldn't change it. But I do wonder for a linter like this whether we get most of the benefit out of a git-grep based linter like some of the others in `pkg/testutils/lint_test.go`, which I only mention because I imagine it would be a lot faster.

That said, this lays the groundwork for adding more checks in the future around aspects of cookie creation if we want.

dhartunian replied:
I'm just learning about the linter infra so I'll give this a quick shot and see if I get anywhere.

dhartunian replied:
git grep linter is way simpler so I'm keeping it. thanks for the pointer.

stevendanna commented on pkg/testutils/lint/passes/norawcookies/testdata/src/a/a.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+func init() {
+	_ = http.Cookie{ // want `raw cookie construction not permitted!`
+		Name: "shouldn't be allowed!",
+	}
```
Out of pure curiosity, does this linter catch an aliased import?

dhartunian replied:
Nope, but I was able to add it.

stevendanna reviewed (approved):
Thanks for going the extra mile here.

I left some comments but neither are blocking.

dhartunian committed:
lint: add linter to reject raw http.Cookie structs

We previously introduced bugs when creating HTTP cookies in new code
that did not have the proper fields set. This linter aims to enforce
use of cookies only through pre-generated constructors in `pkg/server/
authserver/cookie.go`. This will reduce the chance of errors in the
future and keep things consistent.

Epic: None
Release note: None

abarganier reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119748#-:-NsJh7es98_NwKsT1148:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @stevendanna, and @xinhaoz)

___
*[`pkg/server/server_controller_http.go` line 101 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/119748#-NsJfoTQ-JppxkQmshys:-NsJfoTQ-JppxkQmshyt:b-ndu0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbcd1203db5edd5de53a18cd22bd710c8cfe07d3/pkg/server/server_controller_http.go#L101)):*
> ```Go
> 		// Clear session and tenant cookies since it appears they reference invalid state.
> 		http.SetCookie(w, authserver.CreateEmptySessionCookieWithImmediateExpiry(!c.disableTLSForHTTP))
> 		http.SetCookie(w, authserver.CreateEmptyTenantCookieWithImmediateExpiry(!c.disableTLSForHTTP))
> ```

Nice! This is a big win for readability. Good idea.



_Code quote:_
```Go
authserver.CreateEmptyTenantCookieWithImmediateExpiry
```
___
*[`pkg/server/authserver/cookie.go` line 214 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/119748#-NsJgIW10c7r_IWGD_f3:-NsJgIW22ovtp6Ik52J_:b1z41mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbcd1203db5edd5de53a18cd22bd710c8cfe07d3/pkg/server/authserver/cookie.go#L214)):*
> ```Go
> // be optionally marked `Secure` if you're runinning in secure mode by
> // setting the `forHttpsOnly` argument to true.
> func CreateTenantCookie(value string, forHTTPSOnly bool) *http.Cookie {
> ```

nit: Here and below, would `TenantSelect` instead of just `Tenant` be a bit clearer?



_Code quote:_
```Go
CreateTenantCookie
```


<!-- Sent from Reviewable.io -->

dhartunian commented:
TFTRs

bors r+
